### PR TITLE
P2450 C++ Standard Library Issues to be moved in Virtual Plenary, Oct. 2021

### LINKED ISSUE #4972: [2021-10 LWG Motion 1] C++ Standard Library Issues to be moved in Virtual Plenary, Oct. 2021
https://wiki.edg.com/pub/Wg21virtual2021-10/StrawPolls/p2450r0.html

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #4972
Fixes cplusplus/papers#1120

### CODE REVIEW

File: source/time.tex [hunk 1]
```diff
@@ -10975,18 +10989,33 @@
 let \tcode{I} be \tcode{basic_istream<charT, traits>\&},
 where \tcode{charT} and \tcode{traits} are template parameters in that context.
 
+\pnum
+\recommended
+Implementations should make it difficult to accidentally store or use
+a manipulator that may contain a dangling reference to a format string,
+for example by making the manipulators produced by parse immovable and
```
Comment (timsong-cpp) [inline source/time.tex:10996]:
```suggestion
for example by making the manipulators produced by \tcode{parse} immovable and
```

File: source/regex.tex [hunk 2]
```diff
@@ -2076,6 +2083,17 @@
 \tcode{size()} returns \tcode{0}.
 \end{itemdescr}
 
+\indexlibraryctor{match_results}%
+\begin{itemdecl}
+explicit match_results(const match_results& m);
```
Comment (timsong-cpp) [inline source/regex.tex:2088]:
```suggestion
match_results(const match_results& m);
```

File: source/locales.tex [hunk 3]
```diff
@@ -2290,6 +2290,22 @@
...
+\begin{itemize}
+\item
+if the conversion specifier returned by Stage 1 is \tcode{\%d},
+\tcode{"0"} is accumulated;
+\item
+if the conversion specifier returned by Stage 1 is \tcode{\%i},
+\tcode{"0x1a"} are accumulated;
+\item
+if the conversion specifier returned by Stage 1 is \tcode{\%g},
+"\tcode{0x1a.bp+07"} are accumulated.
```
Comment (timsong-cpp) [inline source/locales.tex:2304]:
```suggestion
\tcode{"0x1a.bp+07"} are accumulated.
```

File: source/utilities.tex [hunk 4]
```diff
@@ -1464,11 +1464,32 @@
 \end{itemdecl}
 
 \begin{itemdescr}
+\pnum
+Let \exposid{disambiguating-constraint} be:
+\begin{itemize}
+\item
+\tcode{negation<is_same<remove_cvref_t<U0>, tuple>>}
```
Comment (timsong-cpp) [inline source/utilities.tex:1471]:
```suggestion
\tcode{negation<is_same<remove_cvref_t<$\tcode{U}_0$>, tuple>>}
```

File: source/utilities.tex [hunk 5]
```diff
@@ -1464,11 +1464,32 @@
...
+if \tcode{sizeof...(Types)} is 1;
+\item
+otherwise,
+\tcode{bool_constant<!is_same_v<remove_cvref_t<U0>, allocator_arg_t> ||
+is_-\newline{}same_v<remove_cvref_t<T0>, allocator_arg_t>>}
```
Comment (timsong-cpp) [inline source/utilities.tex:1476]:
```suggestion
\tcode{bool_constant<!is_same_v<remove_cvref_t<$\tcode{U}_0$>, allocator_arg_t> ||
is_-\newline{}same_v<remove_cvref_t<$\tcode{T}_0$>, allocator_arg_t>>}
```

File: source/utilities.tex [hunk 6]
```diff
@@ -1464,11 +1464,32 @@
...
-\tcode{sizeof...(Types)} equals \tcode{sizeof...(UTypes)} and
-$\tcode{sizeof...(Types)} \geq 1$ and
-\tcode{is_constructible_v<$\tcode{T}_i$, $\tcode{U}_i$>} is \tcode{true} for all $i$.
+\begin{itemize}
+\item
+\tcode{sizeof...(Types)} equals \tcode{sizeof...(UTypes)},
+\item
+$\tcode{sizeof...(Types)} \geq 1$, and
+\item
+\tcode{conjunction_v\exposid{<disambiguating-constraint},
```
Comment (timsong-cpp) [inline source/utilities.tex:1490]:
```suggestion
\tcode{conjunction_v<\exposid{disambiguating-constraint},
```

File: source/ranges.tex [hunk 7]
```diff
@@ -1066,16 +1066,18 @@
...
 \begin{itemdescr}
 \pnum
-Given an expression \tcode{E} such that \tcode{decltype((E))} is \tcode{T},
+Let \tcode{U} be \tcode{remove_reference_t<T>}
+if \tcode{T} is an rvalue reference type, and \tcode{T} otherwise.
+Given a variable \tcode{u} of type \tcode{U},
 \tcode{T} models \libconcept{borrowed_range} only if
-the validity of iterators obtained from the object denoted by \tcode{E}
-is not tied to the lifetime of that object.
+the validity of iterators obtained from \tcode{U}
```
Comment (timsong-cpp) [inline source/ranges.tex:1073]:
```suggestion
the validity of iterators obtained from \tcode{u}
```

File: source/algorithms.tex [hunk 8]
```diff
@@ -3939,17 +3939,24 @@
...
-  pairwise model \libconcept{sized_sentinel_for}\iref{iterator.concept.sizedsentinel}
-  for the overloads in namespace \tcode{ranges},
+ the types of \tcode{first1}, \tcode{last1}, \tcode{first2}, and \tcode{last2}
+ pairwise model \libconcept{sized_sentinel_for}\iref{iterator.concept.sizedsentinel}
+  and \tcode{last1 - first1 != last2 - first2}
+  for the first overload in namespace \tcode{ranges},
+\item
+   \tcode{R1} and \tcode{R2} each model \libconcept{sized_range} and
+   \tcode{ranges::distance(r1) != ranges::distance(r2)}
+   for the second overload in namespace ranges,
```
Comment (timsong-cpp) [inline source/algorithms.tex:3958]:
```suggestion
   for the second overload in namespace \tcode{ranges},
```

File: source/iostreams.tex [hunk 9]
```diff
@@ -12215,7 +12215,8 @@
 \begin{itemdescr}
 \pnum
 \effects
-Calls \tcode{sb.emit()}.
+Behaves as an unformatted output function\iref{ostream.unformatted}.
+After constructing a sentry object, calls \tcode{sb.emit()}.
```
Comment (timsong-cpp) [inline source/iostreams.tex:12222]: The issue has `sentry` in code font but it looks like "After constructing a sentry object" is commonly in non-code font in the existing text. The slightly more general "sentry object" has both. Should probably be a separate editorial issue.

File: source/iostreams.tex [hunk 10]
```diff
@@ -6892,7 +6892,10 @@
 Then, if \tcode{os.rdbuf()} is a
 \tcode{basic_syncbuf<charT, traits, Allocator>*},
 called \tcode{buf} for the purpose of exposition,
-calls \tcode{buf->emit()}.
+behaves as an unformatted output function\iref{ostream.unformatted}
+of \tcode{os}.
+After constructing a \tcode{sentry} object, calls \tcode{buf->emit()}.
```
Comment (timsong-cpp) [inline source/iostreams.tex:6897]: ...but we should probably be consistent in this PR?

File: source/iterators.tex [hunk 11]
```diff
@@ -5125,7 +5125,7 @@
 \indexlibrarymember{iter_swap}{common_iterator}%
 \begin{itemdecl}
 template<@\libconcept{indirectly_swappable}@<I> I2, class S2>
-  friend void iter_swap(const common_iterator& x, const common_iterator<I2, S2>& y)
+  friend void constexpr iter_swap(const common_iterator& x, const common_iterator<I2, S2>& y)
```
Comment (timsong-cpp) [inline source/iterators.tex:5128]:
```suggestion
  friend constexpr void iter_swap(const common_iterator& x, const common_iterator<I2, S2>& y)
```

File: source/ranges.tex [hunk 12]
```diff
@@ -4477,15 +4490,33 @@
...
-except that \tcode{E} is evaluated only once.
+
+\item
+otherwise, if \tcode{T} is
+a specialization of \tcode{ranges::iota_view}\iref{range.iota.view}
+that models \libconcept{random_access_range} and \libconcept{sized_range},
+then
+\tcode{ranges::iota_view(*ranges::begin(E),
+*(ranges::begin(E) + std::\linebreak{}min<D>(ranges::distance(E), F)))},
+except that \tcode{E} and \tcode{F} are each evaluated only once;
```
Comment (timsong-cpp) [inline source/ranges.tex:4519]: `F` is only mentioned once in the expression, so clearly it's evaluated only once and we don't need to say it again. I think this was a copy-paste error of mine, sorry :(

```suggestion
except that \tcode{E} is evaluated only once;
```

File: source/iostreams.tex (hunk 10 above)
Comment (jensmaurer) [reply source/iostreams.tex:6897]: Yeah, maybe. I've created #5031 for the general fix, but given the direction there, I'd like to keep this as-is.

File: source/iterators.tex [hunk 13]
```diff
@@ -2935,24 +2938,31 @@
...
 
 \pnum
 \effects
-If \tcode{S} and \tcode{I} model \tcode{\libconcept{sized_sentinel_for}<S, I>},
-returns \tcode{(last - first)};
-otherwise, increments
-\tcode{first} until \tcode{last} is reached and returns
-the number of increments.
+Increments \tcode{first} until \tcode{last} is reached and
+returns the number of increments.
```
[the remark this answers is not in the document]
Comment (timsong-cpp) [reply source/iterators.tex:2953]: Yes, see https://lists.isocpp.org/lib/2021/10/20762.php

**Pull request merged**